apps/prairielearn/src/lib/editors.js

```diff
@@ -82,132 +83,142 @@ export class Editor {
-                //
-                // If pushing errored, the reset will get us back to a known good state.
-                await cleanAndResetRepository(this.course, gitEnv, job);
-              }
+        if (!config.fileEditorUseGit) {
+          await syncCourseFromDisk(this.course, startGitHash, job);
+          return;
+        }
 
-              job.data.syncAttempted = true;
+        try {
+          await job.exec('git', ['add', ...this.pathsToAdd], {
+            cwd: this.course.path,
+            env: gitEnv,
+          });
+          await job.exec(
+            'git',
+            [
+              '-c',
+              `user.name="${this.user.name}"`,
+              '-c',
+              `user.email="${this.user.uid}"`,
+              'commit',
+              '-m',
+              this.commitMessage,
+            ],
+            {
+              cwd: this.course.path,
+              env: gitEnv,
+            },
+          );
+        } catch (err) {
+          await cleanAndResetRepository(this.course, gitEnv, job);
+          throw err;
+        }
 
-              await syncCourseFromDisk(this.course, startGitHash, job);
+        try {
+          job.data.pushAttempted = true;
 
-              // As with `job.data.pushSucceeded` above, we'll check this flag
-              // on the `editError` page to know if syncing failed so we can
-              // display appropriate instructions.
-              job.data.syncSucceeded = true;
-            });
+          await job.exec('git', ['push'], {
+            cwd: this.course.path,
+            env: gitEnv,
           });
-        },
-      ],
-      (err) => {
-        callback(err, jobSequenceId);
-      },
-    );
+
+          // We'll look for this flag on the `editError` page to know if
+          // we need to display instructions to recover from a failed push.
+          job.data.pushSucceeded = true;
+        } finally {
+          // Regardless of whether we error, we'll do a clean and reset:
+          //
+          // If pushing succeeded, the clean will remove any empty directories
+          // that might have been left behind by operations like renames.
+          //
+          // If pushing errored, the reset will get us back to a known good state.
+          await cleanAndResetRepository(this.course, gitEnv, job);
+        }
+
+        job.data.syncAttempted = true;
+
+        await syncCourseFromDisk(this.course, startGitHash, job);
+
+        // As with `job.data.pushSucceeded` above, we'll check this flag
+        // on the `editError` page to know if syncing failed so we can
+        // display appropriate instructions.
+        job.data.syncSucceeded = true;
+      });
+    });
+    return jobSequenceId;
   }
 
+  doEdit = callbackify(async () => {
+    const serverJob = await this.setupEditAsync();
+    const jobSequenceId = await this.doEditAsync(serverJob);
+    return jobSequenceId;
+  });
```

Comment:
This won't do what you want it to. I'd encourage you to add ` throw new Error('testing');` to `doEditAsync` and then try to change an assessment QID from the assessment settings page. Note how you're redirected to something like `/pl/course_instance/2/instructor/edit_error/undefined? That's because `callbackify` will only ever call the callback with `(err)` or `(null, jobSequenceId)`, never `(err, jobSequenceId)` like it used to.

You'll need to write this manually without `callbackify`. It'll end up looking something like this:

```
doEdit(callback) {
  this.setupEditAsync().then(
    (serverJob) => {
      this.doEditAsync(serverJob).then(
        () => callback(null, serverJob.jobSequenceId),
        (err) => callback(err, serverJob.jobSequenceId)
      );
    },
    (err) => {
      callback(err);
    }
  );
}
```

Note how in accordance with my other comment, I didn't use a `jobSequenceId` returned from `doEditAsync`.

Reply:
This has now been updated to the above code where we no longer use `callbackify`.